libraries/AP_Relay/AP_Relay.cpp

```diff
@@ -128,61 +120,237 @@ AP_Relay::AP_Relay(void)
     }
-    const uint32_t now = AP_HAL::millis();
-    _pin_states = value ? _pin_states | (1U<<instance) : _pin_states & ~(1U<<instance);
-    if ((_pin_states != _last_logged_pin_states) && (now - _last_log_ms > 10)) {
-       AP::logger().Write("RELY", "TimeUS,State", "s-", "F-", "QB",
-                                        AP_HAL::micros64(),
-                                        _pin_states);
-       _last_log_ms = now;
-       _last_logged_pin_states = _pin_states;
+
+    for (uint8_t instance = 0; instance < AP_RELAY_NUM_RELAYS; instance++) {
+        if (function != _params[instance].function) {
+            continue;
+        }
+
+        set_pin_by_instance(instance, value);
+    }
+}
+
+// set a pins output state by instance and log if required
+// this is an internal helper, instance must have already been validated to be in range
+void AP_Relay::set_pin_by_instance(uint8_t instance, bool value) {
+    const int8_t pin = _params[instance].pin;
+    if (pin == -1) {
+        // no valid pin to set it on, skip it
+        return;
     }
+
 #if AP_SIM_ENABLED
     if (!(AP::sitl()->on_hardware_relay_enable_mask & (1U << instance))) {
         return;
     }
 #endif
-    hal.gpio->pinMode(_pin[instance], HAL_GPIO_OUTPUT);
-    hal.gpio->write(_pin[instance], value);
+
+    hal.gpio->pinMode(pin, HAL_GPIO_OUTPUT);
+    const bool initial_value = (bool)hal.gpio->read(pin);
+
+    if (initial_value != value) {
+        hal.gpio->write(pin, value);
+        AP::logger().Write("RELY", "TimeUS,Instance,State", "s#-", "F--", "QBB",
+                            AP_HAL::micros64(),
+                            instance,
+                            value);
+    }
+}
+
+void AP_Relay::set(const uint8_t instance, const bool value)
+{
+    if (instance >= AP_RELAY_NUM_RELAYS) {
+        return;
+    }
+
+    if (_params[instance].function != AP_Relay_Params::FUNCTION::RELAY) {
+        return;
+    }
+
+    set_pin_by_instance(instance, value);
 }
 
 void AP_Relay::toggle(uint8_t instance)
 {
-    if (instance < AP_RELAY_NUM_RELAYS && _pin[instance] != -1) {
-        bool ison = hal.gpio->read(_pin[instance]);
-        set(instance, !ison);
+    if (instance < AP_RELAY_NUM_RELAYS) {
+        set(instance, !get(instance));
     }
 }
 
 // check settings are valid
 bool AP_Relay::arming_checks(size_t buflen, char *buffer) const
 {
     for (uint8_t i=0; i<AP_RELAY_NUM_RELAYS; i++) {
-        int8_t pin = _pin[i].get();
+        const int8_t pin = _params[i].pin.get();
         if (pin != -1 && !hal.gpio->valid_pin(pin)) {
-            char param_name_buf[11] = "RELAY_PIN";
-            if (i > 0) {
-                hal.util->snprintf(param_name_buf, ARRAY_SIZE(param_name_buf), "RELAY_PIN%u", unsigned(i+1));
-            }
+            // if this assert triggers the param_name_buf probably just needs to be expanded
+            static_assert ((AP_RELAY_NUM_RELAYS > 0) && (AP_RELAY_NUM_RELAYS < 10), "To many relays for the arming message buffer");
```

Comment:
Or, you know, you could just make the thing 14 bytes on the stack :-)

Reply:
Yep, done.